crates/ruma-state-res/src/event_auth.rs

```diff
@@ -455,7 +503,20 @@ fn valid_membership_change(
                 let allow = join_rules == JoinRule::Invite
                     && (target_user_current_membership == MembershipState::Join
                         || target_user_current_membership == MembershipState::Invite)
-                    || join_rules == JoinRule::Public;
+                    || join_rules == JoinRule::Public
+                    || (room_version.restricted_join_rules
+                        // 0. room version of 8 ^ and join rule of restricted
+                        && restricted
+                        // 1. The user's previous membership was invite or join
+                        && matches!(
+                            target_user_current_membership,
+                            MembershipState::Invite | MembershipState::Join
+                        ))
+                        // TODO: implement this better the function is awful lots of code dup
+                        //
+                        // 2. The join event has a valid signature from a homeserver whose
+                        // users have the power to issue invites.
+                        && restricted_join_rules_auth();
```

Comment:
The indentation is kind of misleading here, I think `allow` should be split into named (as in local variables) sub-conditions. Also I don't get why there's a closure being passed here.

Reply:
Ugh, I thought I commented on this a long time ago.

It's a closure to avoid having to pass the `fetch_state` closure, `content`, `auth_user` and maybe one other ID or event. I know this isn't ideal but I was unsure of how else to add all the complexity for a "cold" path that would become hot eventually.

Reply:
I think it makes more sense to just pass these things. Maybe some further refactoring can be done later.

Reply:
Should I also not worry about the amount of upfront work I end up doing (calling `fetch_state` a few times) even though I may not use any of them if the event isn't the type I have to check (if it's not a member event of a room >= 8?

Reply:
If you need intermediate bindings, I'd rather introduce a little bit of nesting than doing unnecessary upfront work.